File: model/src/main/java/com/bnorm/barkeep/model/Book.java
package com.bnorm.barkeep.model;

import java.util.Set;

import javax.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonView;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
public interface Book extends HasId, Comparable<Book> {

  @JsonView({Book.class, User.class})
  @Nullable
  String getTitle();

  @JsonView({Book.class, User.class})
  @Nullable
  String getDescription();

  @JsonView({Book.class})
  @Nullable
  User getOwner();

  @JsonView(Book.class)
  Set<Recipe> getRecipes();

  @Override
  default int compareTo(Book o) {
    return COMPARATOR.compare(this, o);
  }
}

File: model/src/main/kotlin/com/bnorm/barkeep/model/Book.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonView

interface BookSpec {

  @get:JsonView(Book::class, User::class)
  val title: String?

  @get:JsonView(Book::class, User::class)
  val description: String?

  @get:JsonView(Book::class)
  val owner: User?
}

@JsonInclude(JsonInclude.Include.NON_EMPTY)
interface Book : BookSpec, HasId, Comparable<Book> {

  @get:JsonView(Book::class)
  val recipes: Set<Recipe>?

  override fun compareTo(other: Book): Int {
    return HasId.COMPARATOR.compare(this, other)
  }
}
